include/Validation.h

```diff
@@ -122,25 +126,26 @@ using ValidateFieldArguments = std::map<std::string_view, ValidateArgumentValueP
 
 struct ValidateField
 {
-	ValidateField(std::string&& returnType, std::optional<std::string_view> objectType,
-		std::string_view fieldName, ValidateFieldArguments&& arguments);
+	ValidateField(ValidateType&& returnType, ValidateType&& objectType, std::string_view fieldName,
+		ValidateFieldArguments&& arguments);
 
 	bool operator==(const ValidateField& other) const;
 
-	std::string returnType;
-	std::optional<std::string_view> objectType;
+	ValidateType returnType;
+	ValidateType objectType;
 	std::string_view fieldName;
 	ValidateFieldArguments arguments;
 };
 
-using ValidateTypeKinds = std::map<std::string_view, introspection::TypeKind>;
+using ValidateTypes = std::map<std::string_view, ValidateType>;
```

Comment:
any reason to keep using map instead of unordered_map?

Reply:
I'm acutally thinking of replacing them all with sorted vectors: https://youtu.be/fHNmRkzxHWs?t=2818

Reply:
The key with vectors is to avoid reallocating on append by reserving or resizing to the exact size you need, and not  inserting/removing other than at the end. But for most of these objects that should be doable, they're static for the lifetime of the service.

Reply:
if you're moving to sorted and looking into static... why not target `array`? There is a generator tool to help that, then we can try to get more stuff into `.ro-data` section of the binaries.

There is no need to get the `vector` helpers.

The binary search is not as fast as an unordered map, but I think it doesn't matter much for these use cases. If so we can look into integrating https://www.gnu.org/software/gperf/manual/gperf.html or an equivalent later.

Reply:
> if you're moving to sorted and looking into static... why not target `array`? There is a generator tool to help that, then we can try to get more stuff into `.ro-data` section of the binaries.

Most of them still have a size which varies, so it couldn't be built into the `graphqlservice` library directly.  The generated code would need to pass a span/range into `graphqlservice` to get dynamic sizing backed by `std::array` with a fixed size in the generated code.

C++20 adds span/range, so that would make sense whenever cppgraphqlgen moves to that language version (I'm tentatively planning on doing that for v4.0). There's also the `gsl` (C++ Core Guideline Support Library) which adds lookalike definitions of `std::span<>`, so that's another option that makes C++20 adoption easier down the road. But I'm reluctant to add another build dependency unless/until it's absolutely worth it.

> The binary search is not as fast as an unordered map, but I think it doesn't matter much for these use cases. If so we can look into integrating https://www.gnu.org/software/gperf/manual/gperf.html or an equivalent later.

I'm not so worried about the quality of the `std::hash` implementation for `std::string_view`, and integers/pointers can just hash to their own value.

The multiple levels of indirection and poor data locality (cache misses) in `std::unordered_map` overwhelm the savings on lookup compared to a binary search, until you get to a very large number of items. Anecdotal benchmarks I've heard say you need at least 100 entries to make `std::unordered_map` break even, but really we should do some of our own benchmarking/profiling.

You mentioned you are dealing with large numbers of fields per object, so maybe you're more likely to hit that break even point than I expected. If you want to try comparing them for your scenario, take a look at https://github.com/microsoft/cppgraphqlgen/pull/121/commits/fa772a1076ba08779b1d9b3b1ddea6edafcc20d1 where I already did this for the `ResolverMap` generated for `Object`. If you locally change where it declares the `ResolverMap` type as a `std::vector<std::pair<>>` back to `std::unordered_map` and change the lookup in `SelectionVisitor::visitField` back to a call to `find`, you can see if/what difference it makes.